# Introducción del archivo

Este archivo va a servir como ejemplo de cómo importar los datos financieros de sólo una empresa, en este caso, se escoge la empresa ACS dentro del IBEX 35.

AVISO: ESTE ARCHIVO NO DEBE EJECUTARSE, YA QUE SOLO SIRVE DE EJEMPLO PARA IMPORTAR LOS DATOS FINANCIEROS DE UNA EMPRESA EN PARTICULAR. EL ARCHIVO QUE SE DEBE TENER EN CUENTA ES EL SIGUIENTE:

1- Extraccion de datos de yahoo finance empresas IBEX35.ipynb

A lo largo de este archivo, se va a seguir los siguientes pasos:

En primer lugar, empezaremos importando gracias a la API de Yahoo Finance los datos de la empresa ACS

En segundo lugar, especificaremos el intervalo temporal que queremos escoger para analizar.

En tercer lugar, vamos a analizar la información resultante del dataset de ACS

Finalmente, guardaremos en formato .csv el dataset de ACS.

A lo largo de este archivo, se ha separado en 6 principales pasos los procesos anteriores:
Para la extracción de datos de ACS, mostramos los pasos.

# Indice de pasos:

- [Paso 1: Instalar las librerias necesarias](#paso-1-instalar-las-librerias-necesarias)
- [Paso 2: Importar las librerias necesarias para nuestro analisis](#paso-2-importar-las-librerias-necesarias-para-nuestro-analisis)
- [Paso 3: Lista de empresas del IBEX 35](#paso-3-lista-de-empresas-del-ibex-35)
- [Paso 4: Extraccion de dataset de ACS dentro del IBEX 35 a partir de Yahoo Finance](#paso-4-extraccion-de-dataset-de-acs-dentro-del-ibex-35-a-partir-de-yahoo-finance)
    - [Paso 4.1- Obtencion de datos financieros de ACS a partir de Yahoo Finance](#paso-41--obtencion-de-datos-financieros-de-acs-a-partir-de-yahoo-finance)
    - [Paso 4.2- Especificar los parametros de fechas a escoger](#paso-42--especificar-los-parametros-de-fechas-a-escoger)
    - [Paso 4.3- Establecer el formato de fecha](#paso-43--establecer-el-formato-de-fecha)
- [Paso 5: Informacion acerca del dataset](#paso-5-informacion-acerca-del-dataset)
    - [Paso 5.1- Filas y columnas de cada empresa](#paso-51--filas-y-columnas-de-cada-empresa)
    - [Paso 5.2- Variables que vienen en los datos de las empresas a partir de Yahoo Finance](#paso-52--variables-que-vienen-en-los-datos-de-las-empresas-a-partir-de-yahoo-finance)
    - [Paso 5.3- Informacion de las columnas de los datasets de las empresas](#paso-53--informacion-de-las-columnas-de-los-datasets-de-las-empresas)
- [Paso 6: Guardar los datos ACS a partir de Yahoo Finance en archivo CSV](#paso-6-guardar-los-datos-acs-a-partir-de-yahoo-finance-en-archivo-csv)

# Paso 1: Instalar las librerias necesarias

In [1]:
#pip install pandas
#pip install yfinance

# Paso 2: Importar las librerias necesarias para nuestro analisis

In [2]:
import yfinance as yf

import pandas as pd
import datetime

# Paso 3: Lista de empresas del IBEX 35

La empresa ACS cotiza dentro del IBEX 35. El Ibex 35 es el índice bursátil de referencia de la bolsa española conformado por los 35 valores con más liquidez que cotizan en las cuatro bolsas españolas: Madrid, Barcelona, Bilbao y Valencia. 

A continuación, se van a mencionar por orden alfabético el listado de empresas del IBEX 35.

Esta lista está elaborada con el listado de empresas a fecha de 23/04/2024:

|Número| Compañía | Etiqueta |
|-|----------|----------|
|1| Acciona    | ANA.MC   |
|2| Acciona Energías    | ANE.MC   |
|3| ACS    | ACS.MC  |
|4| Acerinox (NO)    | ACX.MC  |
|5| Aena				|AENA.MC
|6|Amadeus			|AMS.MC
|7|ArcelorMittal		|MTS.MC
|8|Banco Sabadell		|SAB.MC
|9|Banco Santander	|SAN.MC
|10|Bankinter			|BKT.MC
|11|BBVA				|BBVA.MC
|12|CaixaBank			|CABK.MC
|13|Cellnex			|CLNX.MC
|14|Colonial			|COL.MC
|15|Enagás				|ENG.MC
|16|Endesa				|ELE.MC
|17|Ferrovial			|FER.MC
|18|Fluidra			|FDR.MC
|19|Grifols			|GRF.MC
|20|IAG				|IAG.MC
|21|Iberdrola			|IBE.MC
|22|Indra (NO)				|IDR.MC
|23|Inditex			|ITX.MC
|24|Logista 			|LOG.MC
|25|Mapfre				|MAP.MC
|26| Meliá				|MEL.MC
|27|Merlin Properties	|MRL.MC
|28|Naturgy			|NTGY.MC
|29|Redeia				|RED.MC
|30|Repsol				|REP.MC
|31|Rovi				|ROVI.MC
|32|Sacyr				|SCYR.MC
|33|Solaria			|SLR.MC
|34|Telefónica			|TEF.MC
|35|Unicaja			|UNI.MC

# Paso 4: Extraccion de dataset de ACS dentro del IBEX 35 a partir de Yahoo Finance

Vamos a extraer mediante la librería de la página web de Yahoo Finance.

Para ello, vamos a importar la librería de Yahoo Finance. A continuación, vamos a especificar de nuevo la librería a pesar de haberse importado en el Paso 1.

In [3]:
import yfinance as yf

Tras tener importada la librería de Yahoo Finance, cabe la posibilidad de acceder entre otros a los datos financieros de ACS o de las empresas del IBEX 35.

Tal y como se ha señalado previamente, empezaremos por el caso de la empresa ACS de aquí en adelante.

En principio, tendremos en cuenta los datos en los días operativos entre las fechas de: 01/01/2019 y 30/04/2024. Es decir, 5 años y 4 meses enteros en total.

Hay que tener en cuenta que en la bolsa del IBEX 35, los días dentro de los fines de semana la bolsa está cerrada. En adicción, hay otros tantos días que no son festivos. Por lo que en cada año no habría 365 días.

A continuación, se muestra:
- Paso 4.1: Obtención de datos financieros de las 35 empresas a partir de Yahoo Finance.
- Paso 4.2: Especificar los parámetros de fechas a escoger.
- Paso 4.3: Establecer el formato de fecha.

## Paso 4.1- Obtencion de datos financieros de ACS a partir de Yahoo Finance

Aquí importamos el Ticker de Yahoo Finance para ACS, y le asignamos una variable con su respectivo nombre:

In [4]:
ACS_data = yf.Ticker('ACS.MC')

## Paso 4.2- Especificar los parametros de fechas a escoger

Vamos a tener en cuenta, que el intervalo temporal a tener en cuenta para la elaboración de los análisis, tendrá por defecto las siguientes fechas:
- La Fecha inicial se establece en el 01/01/2019
- La Fecha final se establece en el 30/04/2024

En cualquier caso, estas fechas se pueden modificar en la siguiente celda a gusto del usuario.

In [5]:
import datetime

#1   fecha_inicial, la cual podemos modificar a nuestro antojo
fecha_inicial = '2019-01-01'

#2   fecha_final, la cual podemos modificar a nuestro antojo
fecha_final = '2024-04-30'

Por otra parte, dejamos especificados los parámetros de start y end, con la fecha inicial y final establecida en la celda anterior.

A continuación, mostramos como quedaría el filtrado de las fechas mediante la aplicación para la empresa ACS:

In [6]:
ACS_df = ACS_data.history(start=fecha_inicial, end=fecha_final)
ACS_df

Open       High        Low      Close  \
Date                                                                    
2019-01-02 00:00:00+01:00  22.949240  22.949240  22.233565  22.656153   
2019-01-03 00:00:00+01:00  22.424409  22.921970  22.356250  22.710678   
2019-01-04 00:00:00+01:00  22.969685  23.167350  22.881078  23.051477   
2019-01-07 00:00:00+01:00  23.242322  23.378641  22.921972  23.126450   
2019-01-08 00:00:00+01:00  23.119635  23.603567  23.065109  23.365011   
...                              ...        ...        ...        ...   
2024-04-23 00:00:00+02:00  38.500000  38.900002  38.459999  38.480000   
2024-04-24 00:00:00+02:00  38.520000  38.700001  38.220001  38.360001   
2024-04-25 00:00:00+02:00  38.480000  38.480000  37.360001  37.700001   
2024-04-26 00:00:00+02:00  38.040001  38.080002  37.700001  37.820000   
2024-04-29 00:00:00+02:00  37.799999  38.139999  37.799999  37.980000   

                            Volume  Dividends  Stock Splits  
Date                                                         
2019-01-02 00:00:00+01:00   579957        0.0           0.0  
2019-01-03 00:00:00+01:00   826120        0.0           0.0  
2019-01-04 00:00:00+01:00   962645        0.0           0.0  
2019-01-07 00:00:00+01:00  1501743        0.0           0.0  
2019-01-08 00:00:00+01:00  1264760        0.0           0.0  
...                            ...        ...           ...  
2024-04-23 00:00:00+02:00   391787        0.0           0.0  
2024-04-24 00:00:00+02:00   732447        0.0           0.0  
2024-04-25 00:00:00+02:00   503456        0.0           0.0  
2024-04-26 00:00:00+02:00   522897        0.0           0.0  
2024-04-29 00:00:00+02:00   341569        0.0           0.0  

[1364 rows x 7 columns]

## Paso 4.3- Establecer el formato de fecha

A continuación, vamos a establecer el formato de fecha %Y-%m-%d , que es básicamente el YYYY-MM-DD. 

Este formato no está asociado específicamente a un país en particular, sino que es un estándar de formato de fecha ampliamente utilizado en la programación y en sistemas informáticos en general. Es parte de la especificación de formateo de fecha y hora de la biblioteca estándar de Python y se utiliza en muchas otras herramientas y lenguajes de programación.

A continuación, lo aplicamos al dataframe de ACS:

In [7]:
ACS_df.index = ACS_df.index.strftime('%Y-%m-%d')
ACS_df

Open       High        Low      Close   Volume  Dividends  \
Date                                                                         
2019-01-02  22.949240  22.949240  22.233565  22.656153   579957        0.0   
2019-01-03  22.424409  22.921970  22.356250  22.710678   826120        0.0   
2019-01-04  22.969685  23.167350  22.881078  23.051477   962645        0.0   
2019-01-07  23.242322  23.378641  22.921972  23.126450  1501743        0.0   
2019-01-08  23.119635  23.603567  23.065109  23.365011  1264760        0.0   
...               ...        ...        ...        ...      ...        ...   
2024-04-23  38.500000  38.900002  38.459999  38.480000   391787        0.0   
2024-04-24  38.520000  38.700001  38.220001  38.360001   732447        0.0   
2024-04-25  38.480000  38.480000  37.360001  37.700001   503456        0.0   
2024-04-26  38.040001  38.080002  37.700001  37.820000   522897        0.0   
2024-04-29  37.799999  38.139999  37.799999  37.980000   341569        0.0   

            Stock Splits  
Date                      
2019-01-02           0.0  
2019-01-03           0.0  
2019-01-04           0.0  
2019-01-07           0.0  
2019-01-08           0.0  
...                  ...  
2024-04-23           0.0  
2024-04-24           0.0  
2024-04-25           0.0  
2024-04-26           0.0  
2024-04-29           0.0  

[1364 rows x 7 columns]

# Paso 5: Informacion acerca del dataset

Vamos a analizar la información que nos viene finalmente de ACS.

## Paso 5.1- Filas y columnas de cada empresa
En primer, lugar tenemos las siguientes filas y columnas para ACS:

In [8]:
ACS_df.shape

(1364, 7)

Tenemos un total de:
- 1364 filas
- el index de 'Date' y 7 columnas

In [9]:
ACS_df.head(5)

Open       High        Low      Close   Volume  Dividends  \
Date                                                                         
2019-01-02  22.949240  22.949240  22.233565  22.656153   579957        0.0   
2019-01-03  22.424409  22.921970  22.356250  22.710678   826120        0.0   
2019-01-04  22.969685  23.167350  22.881078  23.051477   962645        0.0   
2019-01-07  23.242322  23.378641  22.921972  23.126450  1501743        0.0   
2019-01-08  23.119635  23.603567  23.065109  23.365011  1264760        0.0   

            Stock Splits  
Date                      
2019-01-02           0.0  
2019-01-03           0.0  
2019-01-04           0.0  
2019-01-07           0.0  
2019-01-08           0.0

## Paso 5.2- Variables que vienen en los datos de las empresas a partir de Yahoo Finance

En segundo lugar, desde Yahoo Finance, estamos importando de cada empresa las siguientes variables:

Como podemos observar, hay 7 columnas que vamos a explicar brevemente.

1- Date (index)

Esta columna la hemos creado previamente. Date(fecha) se refiere al punto específico en el tiempo en el que se lleva a cabo una transacción o una serie de transacciones en el mercado de valores. La fecha se indica generalmente en un formato estándar de día, mes y año y es crucial para el registro preciso de las operaciones financieras.

2- Open

Open (Apertura) describe el precio al que una acción comienza a ser negociada al inicio de un periodo determinado de cotización, como el inicio de una jornada de trading o el inicio de un intervalo de tiempo específico.

3- High

High (Alto) se refiere al precio más alto alcanzado por una acción durante el periodo de tiempo analizado. Representa el punto máximo de valoración alcanzado por la acción en el mercado durante ese periodo.

4- Low

Low (Baja) es el precio más bajo registrado para una acción durante el periodo de tiempo en cuestión. Indica el nivel mínimo de valoración al que la acción ha llegado en el mercado durante ese periodo de tiempo.

5- Close

Close (Cierre) describe el precio al que una acción finaliza su cotización al final de un periodo específico, como el cierre de una jornada de trading o el final de un intervalo de tiempo determinado.

6- Volume

Volume (Volumen) indica el número total de acciones que se han negociado durante un periodo de tiempo específico. El volumen es un indicador clave de la actividad de trading en el mercado de valores y puede proporcionar información sobre el interés y la liquidez de una acción en particular.

7- Dividends

Dividends (Dividendos) indica los dividendos pagados por las acciones de la empresa en la fecha correspondiente, si los hubiera. Los dividendos representan los pagos periódicos realizados a los accionistas como una distribución de las ganancias de la empresa.

8- Stock Splits
 
Stock Splits (División de acciones) indica si se produjeron divisiones de acciones de ACS en la fecha indicada. Un split de acciones implica la división de las acciones existentes en múltiples acciones nuevas, generalmente con el objetivo de ajustar el precio por acción y hacerlas más accesibles para los inversores.

## Paso 5.3- Informacion de las columnas de los datasets de las empresas

En tercer lugar, vamos a obtener más información de las columnas:

In [10]:
ACS_df.columns

Index(['Open', 'High', 'Low', 'Close', 'Volume', 'Dividends', 'Stock Splits'], dtype='object')

In [11]:
ACS_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1364 entries, 2019-01-02 to 2024-04-29
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Open          1364 non-null   float64
 1   High          1364 non-null   float64
 2   Low           1364 non-null   float64
 3   Close         1364 non-null   float64
 4   Volume        1364 non-null   int64  
 5   Dividends     1364 non-null   float64
 6   Stock Splits  1364 non-null   float64
dtypes: float64(6), int64(1)
memory usage: 85.2+ KB


Por otra parte, vamos a comprobar si hay valores nulos o NaN. Primero vemos a simple vista si en general puede haber datos nulos:

In [12]:
ACS_df.isna()

Open   High    Low  Close  Volume  Dividends  Stock Splits
Date                                                                   
2019-01-02  False  False  False  False   False      False         False
2019-01-03  False  False  False  False   False      False         False
2019-01-04  False  False  False  False   False      False         False
2019-01-07  False  False  False  False   False      False         False
2019-01-08  False  False  False  False   False      False         False
...           ...    ...    ...    ...     ...        ...           ...
2024-04-23  False  False  False  False   False      False         False
2024-04-24  False  False  False  False   False      False         False
2024-04-25  False  False  False  False   False      False         False
2024-04-26  False  False  False  False   False      False         False
2024-04-29  False  False  False  False   False      False         False

[1364 rows x 7 columns]

Y a continuación, si los hay a lo largo de todo el dataset...

In [13]:
ACS_df.isna().sum()

Open            0
High            0
Low             0
Close           0
Volume          0
Dividends       0
Stock Splits    0
dtype: int64

Como podemos observar, no hay valores nulos, por lo que no será necesaria una limpieza del dataset.

Podemos confirmar que los datos que provienen de Yahoo Finance vienen de forma relativamente limpia para poder analizarlos posteriormente.

# Paso 6: Guardar los datos ACS a partir de Yahoo Finance en archivo CSV

En estos momentos, vamos a proceder a guardar la información proveniente de Yahoo Finance acerca de ACS en un archivo .csv

In [14]:
#ACS_df.to_csv('../Data/ACS_dataset.csv')

Aquí ya tenemos el primer archivo guardado en su correspondiente repositorio.

Se muestra como anulado el código(#) de la celda previa para no tener que crearse un .csv nuevo cada vez que se ejecuta todo en este archivo

In [15]:
ACS_df

Open       High        Low      Close   Volume  Dividends  \
Date                                                                         
2019-01-02  22.949240  22.949240  22.233565  22.656153   579957        0.0   
2019-01-03  22.424409  22.921970  22.356250  22.710678   826120        0.0   
2019-01-04  22.969685  23.167350  22.881078  23.051477   962645        0.0   
2019-01-07  23.242322  23.378641  22.921972  23.126450  1501743        0.0   
2019-01-08  23.119635  23.603567  23.065109  23.365011  1264760        0.0   
...               ...        ...        ...        ...      ...        ...   
2024-04-23  38.500000  38.900002  38.459999  38.480000   391787        0.0   
2024-04-24  38.520000  38.700001  38.220001  38.360001   732447        0.0   
2024-04-25  38.480000  38.480000  37.360001  37.700001   503456        0.0   
2024-04-26  38.040001  38.080002  37.700001  37.820000   522897        0.0   
2024-04-29  37.799999  38.139999  37.799999  37.980000   341569        0.0   

            Stock Splits  
Date                      
2019-01-02           0.0  
2019-01-03           0.0  
2019-01-04           0.0  
2019-01-07           0.0  
2019-01-08           0.0  
...                  ...  
2024-04-23           0.0  
2024-04-24           0.0  
2024-04-25           0.0  
2024-04-26           0.0  
2024-04-29           0.0  

[1364 rows x 7 columns]